include/fmt/format.h

```diff
@@ -3217,26 +3217,29 @@ template <typename T> inline const void* ptr(const std::shared_ptr<T>& p) {
   return p.get();
 }
 
-template <typename It, typename Char> struct arg_join : internal::view {
-  It begin;
-  It end;
+template <typename It, typename S, typename Char>
+struct arg_join : internal::view {
+  It first;
+  S last;
   basic_string_view<Char> sep;
 
-  arg_join(It b, It e, basic_string_view<Char> s) : begin(b), end(e), sep(s) {}
+  arg_join(It f, S l, basic_string_view<Char> s) : first(f), last(l), sep(s) {}
+  It begin() const { return first; }
+  S end() const { return last; }
```

Comment:
`begin()/end()` accessors are not needed either - we should discourage use `arg_join` as a range.

Reply:
I disagree with this; the object itself is useful, and I can even think of situations where it would be beneficial to be able to use it as a range.

Currently it is a struct, and thus it's internals are in the public API (intended or otherwise); it currently *can* be used as range, and probably is. There is no harm in allowing it to act as one. This change just makes it more standard compiling. (Only suggesting the data to be private)

If you do not intend for the user to "use" it, other than an intermediate type; then please consider making it a class with only private data.

If you do not wish the user to use the structure, why bother having the type, and just return a `std::string`? Or, better still, to allow for different `std::basic_string` types, have implicit conversion operators defined. With the data private.

Reply:
It's public only for historical reasons. We should probably make it internal in the next major release.

Reply:
as it stands, and how it is used; that might be a good shout.